applications/MeshMovingApplication/custom_utilities/explicit_fixed_mesh_ale_utilities.cpp

```diff
@@ -0,0 +1,245 @@
+        auto &r_mesh_disp = it_node->FastGetSolutionStepValue(MESH_DISPLACEMENT);
+        r_mesh_disp = ZeroVector(3);
+
+        // Check if any structure node is found
+        if (n_str_nodes != 0){
+            // Compute the closest structure node MESH_DISPLACEMENT
+            double min_distance = 1.01;
+            Node<3>::Pointer str_node_ptr = nullptr;
+            for(unsigned int i_str = 0; i_str < n_str_nodes; ++i_str) {
+                // Compute the structure point weight according to the kernel function
+                const double normalised_distance = std::sqrt(i_fl_str_dists[i_str]) / mSearchRadius;
+
+                // Check if the point is closer than the current one
+                if (normalised_distance < min_distance) {
+                    min_distance = normalised_distance;
+                    str_node_ptr = i_fl_str_nodes[i_str];
+                }
+            }
+
+            // Current step structure pt. DISPLACEMENT values
+            if (str_node_ptr) {
+                const double weight = this->ComputeKernelValue(min_distance);
+                const auto &r_str_disp_0 = str_node_ptr->FastGetSolutionStepValue(DISPLACEMENT,0);
+                const auto &r_str_disp_1 = str_node_ptr->FastGetSolutionStepValue(DISPLACEMENT,1);
+                const auto str_disp = weight * (r_str_disp_0 - r_str_disp_1);
+                if (!it_node->IsFixed(MESH_DISPLACEMENT_X)) {
+                    r_mesh_disp[0] = str_disp[0];
+                }
+                if (!it_node->IsFixed(MESH_DISPLACEMENT_Y)) {
+                    r_mesh_disp[1] = str_disp[1];
+                }
+                if (!it_node->IsFixed(MESH_DISPLACEMENT_Z)) {
+                    r_mesh_disp[2] = str_disp[2];
+                }
+            }
+        }
+    }
+}
+
+inline double ExplicitFixedMeshALEUtilities::ComputeKernelValue(const double NormalisedDistance){
+    // Epanechnikov (parabolic) kernel function
+    // return (std::abs(NormalisedDistance) < 1.0) ? std::abs((3.0/4.0)*(1.0-std::pow(NormalisedDistance,2))) : 0.0;
+    // Triangle kernel function
+    return (std::abs(NormalisedDistance) < 1.0) ? 1.0 - std::abs(NormalisedDistance) : 0.0;
+}
+
+void ExplicitFixedMeshALEUtilities::CreateVirtualModelPartElements(const ModelPart &rOriginModelPart)
+{
+    // Copy the origin model part elements
+    auto &r_elems = rOriginModelPart.Elements();
+    for(auto &elem : r_elems) {
+        // Set the array of virtual nodes to create the element from the original ids.
+        PointsArrayType nodes_array;
+        auto &r_orig_geom = elem.GetGeometry();
+        for (unsigned int i = 0; i < r_orig_geom.PointsNumber(); ++i){
+            nodes_array.push_back(mrVirtualModelPart.pGetNode(r_orig_geom[i].Id()));
+        }
+
+        // Create the same element but using the virtual model part nodes
+        auto p_elem = elem.Create(elem.Id(), nodes_array, elem.pGetProperties());
+        mrVirtualModelPart.AddElement(p_elem);
```

Comment:
I think it is faster to add the new elements alltogether instead of one-by-one

Reply:
It is. I'll change it. Thanks!

Reply:
Done.